Refactoring NPC movement by getting rid of separate definite movement and indefinite movement structs

bvvill opened the pull request:
As detailed in #1475

bvvill committed:
many a change

bvvill committed:
forgot a couple of crucial includes

bvvill committed:
*productive whistling noises*

bvvill committed:
forgot semicolon, messed up the order of some arguments, blah blah

bvvill committed:
easing on down the road

bvvill committed:
I cross my heart and hope to die

bvvill committed:
probably the only thing I didn't think about checking before commiting anything

bvvill committed:
i think i'll be done soon

bvvill committed:
inch resting

bvvill committed:
okay, I have a thought

bvvill committed:
maybe?

bvvill committed:
can we get this show back on the road?

bvvill committed:
makes sense

bvvill committed:
shooting in the dark again

bvvill committed:
very confusion

bvvill committed:
idk

bvvill committed:
maybe?

bvvill committed:
possibly

bvvill committed:
definitely messing up something with utlist

bvvill committed:
might be getting somewhere

bvvill committed:
I will be happily surprised and disappointed if this works

bvvill committed:
o my god i didn't realize I would create an infinite loop

bvvill committed:
sure

bvvill committed:
adding a test for get_npc_num_rooms

bvvill committed:
making use of time_ray_free in npc_path_dll_free

bvvill committed:
even more so than the last commit

bvvill committed:
adding descriptions to new functions and removing functions from the npc_move header file that don't need to be there

bvvill committed:
editing function descriptions where needed

bvvill committed:
editing literally one word in a function's description lol

bvvill committed:
astyling src/npc/src and tests/npc/

bvvill committed:
forgot I commented this line out

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

Commit:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

sglenn-uchi commented on include/npc/npc_move.h:
```diff
@@ -155,50 +116,34 @@ npc_mov_t *npc_mov_new(npc_mov_enum_t mov_type, char *room_id);
  */
```
Incomplete function header

sglenn-uchi replied:
oh i just expanded it and saw

sglenn-uchi commented on include/npc/npc_move.h:
```diff
@@ -155,50 +116,34 @@ npc_mov_t *npc_mov_new(npc_mov_enum_t mov_type, char *room_id);
 
+// "GET" FUNCTIONS ------------------------------------------------------------
 /*
  * Returns the room that the npc is currently in
```
Incomplete function header (missing parameters)

bvvill replied:
I don't think you're clicking on the arrow to expand the header descriptions

sglenn-uchi commented on include/npc/npc_move.h:
```diff
@@ -224,37 +169,24 @@ char *get_npc_curr_room_id(npc_mov_t *npc_mov);
+unsigned int get_npc_path_direction(npc_mov_t *npc_mov);
 
 /*
  * Gets the number of rooms in an NPC's path
```
Incomplete function header (missing parameters)

sglenn-uchi commented on include/npc/npc_move.h:
```diff
@@ -5,144 +5,105 @@
  */
-npc_mov_t *npc_mov_new(npc_mov_enum_t mov_type, char *room_id);
-
+npc_mov_t *npc_mov_new(npc_mov_enum_t mov_type, char *room_id, double room_time);
```
Return type of should be int if returning SUCCESS/FAILURE or npc_mov_t if returning the struct (contradiction between function description and return value in header comment)

bvvill replied:
Ahhh yes thank you, it's been changed to "Pointer to the new npc_mov_t struct"

sglenn-uchi commented on include/npc/npc_move.h:
```diff
@@ -5,144 +5,105 @@
 
-/* Doubly-Linked List of Room IDs Struct
+/* Struct for containing the start time of a certain phenomenon
+ * and the number of SECONDS that phenomenon is supposed to last,
```
why is seconds all caps only here?

bvvill replied:
to emphasize the change from milliseconds to seconds

sglenn-uchi commented on include/npc/npc_move.h:
```diff
@@ -5,144 +5,105 @@
+    double assigned_time;
+} time_ray_t;
+
+/* Doubly-Linked List of an NPCs path Struct
```
really minor but "path" can be capitalized

sglenn-uchi commented on include/npc/npc_move.h:
```diff
@@ -318,23 +235,12 @@ int move_npc_indefinite(npc_mov_t *npc_mov);
- * 2 successful move to the next room
- * 3 npc has nowhere to move
+ * 0: if move is unsuccessful
+ * 1: if npc has reached the end of the path, reverse_path is called
```
Maybe change redistribute return values such that 1 is the success case and 2 and 3 are the other cases? just thought it might be more intuitive for future programmers to see 0 as failure case and 1 as success case (unless this would result in a lot of other changes to be made)

bvvill replied:
I agree, the change will be pushed in a few minutes

sglenn-uchi commented on include/npc/rooms_npc.h:
```diff
@@ -35,7 +35,6 @@ typedef struct npcs_in_room npcs_in_room_hash_t;
  */
```
Incomplete function headers

bvvill replied:
for which functions ?

sglenn-uchi replied:
oh my bad i think this was before i realized i could expand the function headers

bvvill committed:
correcting npc_mov_new description return value

sglenn-uchi commented on include/npc/rooms_npc.h:
```diff
@@ -35,7 +35,6 @@ typedef struct npcs_in_room npcs_in_room_hash_t;
  */
```
I think the UT_hash_handle component hh was removed so you can remove it from the struct comment

bvvill replied:
wym
```
/* 
 * Struct for adding and handling npcs in rooms 
 * 
 * Components:
 *  room_id: The name of the room
 *  npc_list: Hash table storing the npcs in the room
 *  num_of_npcs: Number of npcs in the room
 */
typedef struct npcs_in_room {
    UT_hash_handle hh;
    char* room_id;
    npc_hash_t *npc_list;
    int num_of_npcs;
} npcs_in_room_t;
```

sglenn-uchi replied:
just looked at this again and i have no idea why i thought that... my bad

sglenn-uchi commented on include/npc/rooms_npc.h:
```diff
@@ -47,7 +46,6 @@ int npcs_in_room_init(npcs_in_room_t *npcs_in_room, char *room_id);
-
 /*
  * Frees resources associated with an npcs_in_room struct
  * Deletes npc_list hashtable, but does not free npcs in hash table
```
Same as earlier since hashtable was removed

bvvill replied:
where was the hashtable removed ?

bvvill committed:
changing output values of move_npc_mov to be more intuitive

sglenn-uchi commented on src/game-state/src/room.c:
```diff
@@ -355,20 +354,17 @@ int npc_one_move_helper(npc_t *npc, npcs_in_room_t *old_npc_room,
+    return SUCCESS;
+}
 
 int npc_one_move(npc_t *npc, room_hash_t *all_rooms)
```
was this defined in a header file? im expanding the code in header files and can't find a definition

bvvill replied:
no , it isn’t ! very weird , i’ll add that when i wakeup later today , nice catch  
!

bvvill replied:
oh nvm , it's actually already defined in `rooms_npc.h`

sglenn-uchi commented on src/npc/src/npc_move.c:
```diff
@@ -340,9 +327,13 @@ int move_npc_indefinite(npc_mov_t *npc_mov)
+        {
+            assert(flip_npc_path_direction(npc_mov) == SUCCESS);
+        }
+        return 2;
```
should 0 be returned here if npc is definite?

bvvill replied:
great observation , and while keeping the return value as 2 for both cases was intentional on my part , I'm convinced that it should be considered an unsuccessful move . Considering the multiple discussions about the outputs of this function , i think it's appropriate to just simplify it to return `SUCCESS` or `FAILURE` . I believe the use of numbers to match specific cases of NPC movement was implemented for easier debugging , and i preserved that style of implementation in my own iterations of the functions , but now i think it's more important to reduce complexity . In order to keep debugging this function relatively easy , I'm going to assign the local variable `path_pos` to be a unique value before every possible return call , to help with tracing .

sglenn-uchi commented on tests/npc/test_npc.c:
```diff
@@ -72,7 +72,7 @@ move_t *create_enemy_moves()
     move_t *head, *earthquake, *poke, *rock_throw;
     head = NULL;
-    earthquake = move_new(1, "earthquake", "", PHYS, NO_TARGET, NO_TARGET, 
+    earthquake = move_new(1, "earthquake", "", PHYS, NO_TARGET, NO_TARGET,
```
move_new is defined in battle_moves.h which isn't included in test_npc.c (it is included in test_battle_ai.c though where create_enemy_moves is defined)

bvvill replied:
tests_npc.c includes npc.h, which includes npc_battle.h, which includes battle_moves.h, so i don't think anything is wrong with this

sglenn-uchi replied:
i just noticed it was included in test_npc_battle.c but not in tests_npc.c but you're right, it won't matter since it's functionally the same

sglenn-uchi commented on tests/npc/test_npc_move.c:
```diff
@@ -325,9 +322,8 @@ Test(npc_mov, auto_gen_movement_definite)
-                 (npc_mov->npc_mov_type.npc_mov_definite->npc_path),
-                 SUCCESS, "delete room_id_dll failed");
+    cr_assert_eq(npc_path_dll_free(npc_mov->path),
+                 SUCCESS, "delete npc_path_dll failed");
```
think "delete" was kept because of old function name
could change to free just for consistency's sake with new function name

bvvill replied:
ah yes , i definitely forgot to change that in this branch, thank u!

sglenn-uchi commented on tests/npc/test_npc_move.c:
```diff
@@ -382,9 +378,9 @@ Test(npc_mov, auto_gen_movement_indefinite)
-                 SUCCESS, "delete room_id_dll failed");
+    cr_assert_eq(npc_path_dll_free
+                 (npc_mov->path),
+                 SUCCESS, "delete npc_path_dll failed");
```
same as above

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
updating return values for move_npc_mov and dealing with that change amplification

bvvill committed:
fixing tests to accommodate recent changes in output parameters of move_npc_mov as well as the abstraction of npc_path_dll_free()

bvvill committed:
lol i put FAILED and not FAILURE

sglenn-uchi commented on tests/npc/test_npc_move.c:
```diff
@@ -183,7 +183,7 @@ Test(npc_mov, move_npc_mov_def)
     int check2 = move_npc_mov(npc_mov);
 
-    cr_assert_eq(check2, 1, "move_npc_def() failed");
+    cr_assert_eq(check2, SUCCESS, "move_npc_def() failed");
```
I think "move_npc_def()" is the outdated function name. It should be just move_npc_mov, I believe? This goes for the tests on lines  186, 189, 194, as well 261.

bvvill replied:
yes you're right , updating that now , thank u !

sglenn-uchi reviewed (approved):
Aside from the final stylistic comments I made here, I think this PR is complete. I think consolidating the definitive and indefinite movement structs streamlines a lot of the implementation and understanding of the design. This PR looks good and is ready for Lily's final review!

bvvill committed:
updating move_npc_def() to just move_npc_mov() in some assertion failure messages in test_npc_move.c

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
adding newlines to the EOF of the header files i've already changed for this issue

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

lilyehsani commented on src/npc/src/npc_move.c:
```diff
@@ -1,58 +1,165 @@
+    strcpy(npc_mov->track, room_id);
+    npc_mov->npc_path_pos = 0;
+    npc_mov->npc_path_direction = NPC_MOV_ORIGINAL;
+    npc_path_dll_t *head = NULL; // these next few lines are VERY important
```
Add a comment explaining why/what they do

lilyehsani commented on src/npc/src/npc_move.c:
```diff
@@ -242,40 +305,17 @@ int flip_npc_path_direction(npc_mov_t *npc_mov)
     if ((current_room->next == NULL) && (current_room->prev == NULL))
-        return 3; // NPC has nowhere to move
+    {
+        path_pos += 1; // For debugging use
```
What do you mean by "For debugging use"? Should these be removed?

bvvill replied:
> great observation , and while keeping the return value as 2 for both cases was intentional on my part , I'm convinced that it should be considered an unsuccessful move . Considering the multiple discussions about the outputs of this function , i think it's appropriate to just simplify it to return `SUCCESS` or `FAILURE` . I believe the use of numbers to match specific cases of NPC movement was implemented for easier debugging , and i preserved that style of implementation in my own iterations of the functions , but now i think it's more important to reduce complexity . In order to keep debugging this function relatively easy , I'm going to assign the local variable `path_pos` to be a unique value before every possible return call , to help with tracing .

so I added those lines in order to aid developers (although I'm probably the only one who will ever have to at this point) who are debugging this function by stepping through it. Since there's only 2 possible return values, but many different ways of reaching then, I figured i could just uniquely change a certain local variable (in this case, path_pos) in order to give something debuggers could set a watchpoint on or something, but yeah thinking about it now these should definitely be deleted, thx !

lilyehsani reviewed (changes requested):
Excellent refactor; this design is much cleaner! Just a couple clarification comments.

bvvill committed:
implementing lily's suggestions

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

lilyehsani committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
fixing errors from merge

bvvill committed:
Merge branch 'dev' into npc/1475-refactoring-npc-movement-by-getting-rid-of-separate-definite-movement-and-indefinite-movement-structs

bvvill committed:
fixing test failure from merge

bvvill committed:
removing unnecessary comment from move_npc_mov function description

lilyehsani reviewed (approved):
Excellent job! Approved.

bvvill commented:
comments from #1475:

Currently, the only difference between the `npc_mov_definite` and `npc_mov_indefinite` structs is that the indefinite one has a hash table of "room times", `npc_room_time_hash_t`, which is just has the room IDs and an integer value for the amount of milliseconds an indefinite NPC is supposed to be in that room. Although a hash_table is technically a doubly-linked list (per uthash's documentation), mapping these time values in the hash table to the steps in the movement path has some problems:
1. If an NPC has a particular room in its path multiple times, it will always stay in that room for the exact same amount of time as before (as opposed to spending 15000 ms in it the first time and 20000 ms the second time, for example), because all values in the hash table must be unique
2. Is unnecessarily complex, and serves as a speed bump in further development of NPC movement

My current plan to alleviate this is to add a time field to the existing `room_id_dll` struct, so that it isn't necessary to have separate structs for definite and indefinite moving NPCs and taking the `mov_type` enum into consideration (where appropriate) will be sufficient, and considerably less complex.

The time field I mentioned in the previous comment will come in the form of the `time_ray_t` struct, which have `time_t start` and `double assigned_time` fields, representing the start time of when an (indefinite) NPC enters a room and the number of seconds that NPC has been assigned to be in that room

This `time_ray_t` struct will be a new member of the `npc_path_dll_t` struct, which I am renaming from `room_id_dll_t`

Definite NPCs will have this `time_ray_t` field to be `NULL`, whereas it will actually be initialized for indefinite NPCs. `init()`, `new`, and `free` functions for the `time_ray_t` struct have been implemented, and they have been abstracted to solely live in npc_move.c, as helper functions.

removed the `move_npc_indefinite` and `move_npc_definite` functions as it's no longer necessary to have separate functions, `move_npc_mov` takes care of both types perfectly

Ran into a little bit of a change amplification problem from going from `int room_time` to `double room_time` in reference to the amount of seconds (which used to be milliseconds, and is being switched in order to be similar to the output of `difftime()` in the standard `time.h` header, which will soon be utilized to its full extent) an indefinite NPC spends in each room along its path, mainly in npc_move.c and room.c

Removed the `register_npc_room_time` function and its test, as `npc_mov_init` now accomplishes the same purpose.

updated `extend_npc_path_indefinite()` and `extend_npc_path_definite()` (and their tests) to reflect changes in the `npc_mov` struct. I've decided to let these functions stay separate because they get called upon often and there's an extra parameter used in `extend_npc_path_indefinite` that would often be unused if the functions were generalized. This makes me really wish there was a good way to do optional arguments in C lol

also, forgot to mention that I've implemented init(), new(), and free() functions for the `npc_path_dll_t` struct, although I've decided to abstract them away to solely reside in `npc_move.c`, since they're basically helper functions

Finally, did some rearranging of the functions in the `npc_move` module so that the functions are sorted by their type of operation, like struct, "get", compare, and "do something" (I could think of no better name, even update didn't totally apply lol) functions


``` strcpy(npc_mov->track, room_id);
    npc_mov->npc_path_pos = 0;
    npc_mov->npc_path_direction = NPC_MOV_ORIGINAL;
    npc_mov->path = NULL;
    npc_path_dll_t *elt = npc_path_dll_new(mov_type, room_id, room_time);
    DL_APPEND(npc_mov->path, elt);
    return SUCCESS;
```

basically all of the failed builds on this branch are because I had set up `npc_mov_init` like the above code, instead of the below. Letting this serve as a painful reminder that initializing the head of a utlist is essential, as sussing this out took me 2 excruciatingly-long hours

```
int npc_mov_init(npc_mov_t *npc_mov, npc_mov_enum_t mov_type, char *room_id,
                 double room_time)
{
    assert(npc_mov != NULL);
    npc_mov->mov_type = mov_type;
    strcpy(npc_mov->track, room_id);
    npc_mov->npc_path_pos = 0;
    npc_mov->npc_path_direction = NPC_MOV_ORIGINAL;
    npc_path_dll_t *head = NULL; // these next few lines are VERY important
    npc_path_dll_t *elt = npc_path_dll_new(mov_type, room_id, room_time);
    DL_APPEND(head, elt);
    npc_mov->path = head;
    return SUCCESS;
}
```

lilyehsani commented:
Issue Score: Excellent

Comments:
Great work on this issue! The summary for this issue is too terse. A good rule of thumb is that you should be writing at least one full paragraph in your summary (see https://uchicago-cs.github.io/cmsc22000/project/issue-pr-grading.html#descriptive-summary for examples).

<!-- {"score": "excellent", "minor": false} -->
